Draft for BookExemplary feature

Zsolt2000 opened the pull request.

Zsolt2000 committed:
Added BookExemplary classes
Added future TODOs

Zsolt2000 committed:
Added Wrapper for NoSuchAlgorithmException
Renamed EntityAccountNotVerifiedException and EntityBadCredentialsException
Implemented GetAllBooks which returns their BookExemplary (code is hot garbage)

Zsolt2000 commented on src/main/java/com/project/bookstore/mapper/BookExemplaryMapper.java:
```diff
@@ -0,0 +1,43 @@
+@Component
+public class BookExemplaryMapper {
+    @Autowired
+    private BookMapper bookMapper;
```
Aici cand vreau sa afisez exemplarele cu cartile lor, am probleme prin faptul ca vreau sa convertesc cu BookMapper, care la randul ei are nevoie de BookExemplaryMapper, creand astfel o dependinta circulara. Aceeasi problema am si cand vreau sa afisez librariile cu cartile lor. O rezolvare pe moment ar fi sa fac un DTO care are doar anumite proprietati, but would like to hear your opinion.

marcopacurariu3 replied:
Corect, dar asta iti ziceam aseara. Nu vad inca un caz in care ai nevoie sa aduci o biblioteca cu toate cartile sale sau o carte cu toate exemplarele sale.

Astfel, cel putin pentru moment, un BookDTO nu va contine o lista de ExemplaryDTO. Si nici un LibraryDTO nu va contine o lista de BookDTO.

Asta pentru ca eu, inca, nu vad un caz in care ai avea nevoie de asa ceva, sa vina datele impreuna, parintele impreuna cu copiii. Tell me if you think I'm wrong.

Zsolt2000 replied:
I mean da, nu prea are rost, cel mult o sa fie sa aducem entitatiile copil DE LA o entitate parinte

Zsolt2000 commented on src/main/java/com/project/bookstore/dto/BookDto.java:
```diff
@@ -17,6 +19,9 @@ public class BookDto {
     private BookLanguage bookLanguage;
     @JsonIgnore
     private LibraryDto libraryDto;
+    private BookExemplaryDto bookExemplary;
```
Codul din clasele BookDto si BookExemplaryDto is hot garbage, but that's not the point. Vreau sa stiu daca abordarea mea e ok. Dupa spusele tale de ieri, la BookExemplaryDto am pus o proprietate care zice cate exemplare sa creeze (setat in RequestBody-ul de la createBook din BookController), iar in BookMapper am facut ca dupa numarul acela sa mapeze atatea Dto-uri de exemplare in entitati. In DB se salveaza corect relatia, dar problema e cu dependinta circulara si , ulterior, cu recursivitatea in Response.

marcopacurariu3 replied:
am raspuns la celalalt comment, sa imi zici daca e clar acum :)

marcopacurariu3 commented on src/main/java/com/project/bookstore/controller/BookController.java:
```diff
@@ -54,9 +52,9 @@ public ResponseEntity<?> getBookById(@PathVariable(name = "bookId") Long bookId)
-    public ResponseEntity<?> getAllBooks(@RequestParam(name = "pageSize") Optional<Integer> pageSize, @RequestParam(name = "pageNumber") Optional<Integer> pageNumber) {
-        if (pageSize.isPresent() && pageNumber.isPresent()) {
-            Page<Book> pageBook = bookService.getAllBooks(PageRequest.of(pageNumber.get(), pageSize.get()));
+    public ResponseEntity<?> getAllBooks(@RequestParam(name = "pageSize",required = false) Integer pageSize, @RequestParam(name = "pageNumber",required = false) Integer pageNumber) {
```
am gasit si comment-ul pe stackoverflow de ziceam ieri: https://stackoverflow.com/questions/26327957/should-java-8-getters-return-optional-type/26328555#26328555

marcopacurariu3 commented on src/main/java/com/project/bookstore/controller/BookExemplaryController.java:
```diff
@@ -0,0 +1,36 @@
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequestMapping("/exemplary")
```
exemplars

marcopacurariu3 commented on src/main/java/com/project/bookstore/dto/BookExemplaryDto.java:
```diff
@@ -0,0 +1,53 @@
+    private String publisher;
+    private int maximumReservationDuration;
+    private BookDto book;
+    //TODO: on Create method have a custom Dto with a nr of exemplars to create
```
corect, ai zis bine

marcopacurariu3 commented on src/main/java/com/project/bookstore/entity/BookExemplary.java:
```diff
@@ -0,0 +1,59 @@
+
+@Entity(name = "exemplary")
+@Table(name = "exemplary", schema = "public")
+public class BookExemplary {
```
nu pot sa zic ca sunt mega fan ca in db e doar exemplary, in @Entity la fel, dar clasa de fapt e BookExemplary. Cred ca i-as fi zis doar Exemplary, dar nu stiu daca mai merita acum sa redenumesti. Intr-adevar, intr-un fel nu e rau sa fie si Book in nume, can't deny.

Zsolt2000 replied:
![Screenshot_20250225_094742_Chrome.jpg](https://github.com/user-attachments/assets/ffb90b42-bd25-4895-8fb1-5401b8f826cc)

Eu am mers pe ideea asta :))))))

marcopacurariu3 commented on src/main/java/com/project/bookstore/entity/BookExemplary.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    public List<BookExemplary> createListWithPredefinedSize(int number) {
+        return IntStream.range(0, number).mapToObj(i -> new BookExemplary()).toList();
```
.map()
.toList()

pe linii separate

marcopacurariu3 commented on src/main/java/com/project/bookstore/entity/BookExemplary.java:
```diff
@@ -0,0 +1,59 @@
+        this.book = book;
+    }
+
+    public List<BookExemplary> createListWithPredefinedSize(int number) {
```
mda... not sure de idee, nu pot sa zic ca imi place foarte mult, cumva ii dai responsabilitatea unui exemplar sa creeze alte exemplare, mi se pare ca nu e responsabilitatea unui exemplar sa faca asta. arguably ar fi al unui book. Dar as lasa logica asta in mapper direct: primesti un DTO care iti zice cate exemplare sa creezi => mapper-ul iti returneaza o lista de exemplare.

marcopacurariu3 commented on src/main/java/com/project/bookstore/exceptions/EntityAccountNotVerifiedException.java:
```diff
@@ -0,0 +1,7 @@
+package com.project.bookstore.exceptions;
+
+public class EntityAccountNotVerifiedException extends RuntimeException {
```
do we need "entity" in numele exceptiei?

Zsolt2000 replied:
Aici am generalizat ca aceeasi exceptie ar fi daca un User si un Librarian nu ar avea contul verificat

marcopacurariu3 replied:
Mda, nu as mai fi pus si "entity" in nume, dar I get it, as you also wish, n-am un strong opinion

marcopacurariu3 commented on src/main/java/com/project/bookstore/exceptions/NoSuchAlgorithmWrapperException.java:
```diff
@@ -0,0 +1,7 @@
+package com.project.bookstore.exceptions;
+
+public class NoSuchAlgorithmWrapperException extends RuntimeException{
```
format

marcopacurariu3 commented on src/main/java/com/project/bookstore/helper/PasswordEncryptor.java:
```diff
@@ -8,17 +10,22 @@ public class PasswordEncryptor {
-            hexString.append(hex);
+            return hexString.toString();
+        } catch (NoSuchAlgorithmException e) {
+            throw new NoSuchAlgorithmWrapperException("Unable to encrypt password", e);
```
cool. sau mai mergea sa fie o exceptie "generala" pentru erori de genul asta la care sa returnezi 500.

marcopacurariu3 replied:
ca sa nu fie, asa, un wrapper atat de specific. Ca oricum exceptia asta generala ar avea un Cause pe care il primeste, astfel ca e un wrapper general. Pe logica faptului ca daca mai apar pe viitor 20 de exceptii, sa nu avem 20 de wrappere.

marcopacurariu3 commented on src/main/java/com/project/bookstore/service/BookExemplaryService.java:
```diff
@@ -0,0 +1,28 @@
+        return bookExemplaryRepository.findAll(pageable);
+    }
+
+    public void deleteExemplaryById(Long id) {
```
acum ma mai gandeam... eu as denumi metodele din service fara sa mai aiba numele entitatilor in ele -> e clar daca daca este in BookExemplary stergi un BookExemplary by id. as lasa stilul JPA, sa nu mai pun numele entitatii

marcopacurariu3 commented on src/main/java/com/project/bookstore/mapper/BookMapper.java:
```diff
@@ -19,6 +23,10 @@ public Book mapBookFromBookDto(BookDto bookDto) {
         book.setNrOfPages(bookDto.getNrOfPages());
         book.setGenre(bookDto.getGenre());
         book.setBookLanguage(bookDto.getBookLanguage());
+        book.setBookExemplars(IntStream.range(0, bookDto.getBookExemplary().getNrOfExemplarsToCreate())
```
hmm, aici ce caz ai facut? cel in care creezi o carte cu tot cu exemplare? inca nu il cerusem, stiu ca am zis ca facem acel "v2" dintre solutii, sa nu cream inca o carte direct cu exemplare.

Zsolt2000 replied:
I mean, exemplarele au legatura stransa cu cartile si de aia am mers pe ideea aia. I mean sure pentru o carte data sa creeam n exemplare ar avea sens

Zsolt2000 committed:
Refactored BookExemplar generation
Renamed BookExemplar classes from BookExemplary and its refferences
Renamed Service methods for similarity to the JPA standard
Added Unit tests for the BookExemplarService class
